File: languages/go/tests/reconcile_test.go
// The reconcile loop (failover-contract §3, runtime-tdd §4) over the mock
// transport: verdict mapping from RAW bodies, degraded-read polling, the
// PermanentRejection escape, policy bounds, and context cancellation.
package tests

import (
	"context"
	"errors"
	"testing"
	"time"

	revaly "github.com/revaly-co/rap-sdk/languages/go"
	"github.com/revaly-co/rap-sdk/languages/go/raptest"
)

func policy(attempts int) revaly.ReconcilePolicy {
	return revaly.ReconcilePolicy{
		MaxAttempts:   attempts,
		OverallBudget: 30 * time.Second,
		InitialDelay:  0, // back-to-back in tests; jitter/backoff has its own test
	}
}

func TestReconcileFindsTerminalOutcomes(t *testing.T) {
	for name, expect := range map[string]struct {
		script  func(*raptest.MockOperation) *raptest.MockOperation
		outcome revaly.TransactionOutcome
	}{
		"approved": {func(o *raptest.MockOperation) *raptest.MockOperation { return o.ReturnsFoundApproved() }, revaly.TransactionOutcomeApproved},
		"declined": {func(o *raptest.MockOperation) *raptest.MockOperation { return o.ReturnsFoundDeclined() }, revaly.TransactionOutcomeDeclined},
	} {
		client, mock := newMockClient(t, nil)
		expect.script(mock.Reconcile("mock-mtx-0001"))
		verdict, err := client.Reconcile(context.Background(), "mock-mtx-0001", policy(3))
		if err != nil {
			t.Fatalf("%s: %v", name, err)
		}
		found, ok := verdict.(*revaly.Found)
		if !ok {
			t.Fatalf("%s: verdict %T, want *Found", name, verdict)
		}
		if found.Outcome != expect.outcome {
			t.Fatalf("%s: outcome %q, want %q", name, found.Outcome, expect.outcome)
		}
		if found.Transaction == nil || found.CorrelationID == "" {
			t.Fatalf("%s: expected an enriched sighting with correlation, got %+v", name, found)
		}
	}
}

func TestReconcileFindsPendingIntent(t *testing.T) {
	client, mock := newMockClient(t, nil)
	mock.Reconcile("mock-mtx-0001").ReturnsPending()
	verdict, err := client.Reconcile(context.Background(), "mock-mtx-0001", policy(2))
	if err != nil {
		t.Fatal(err)
	}
	found, ok := verdict.(*revaly.Found)
	if !ok || found.Outcome != revaly.TransactionOutcomePending || found.Pending == nil {
		t.Fatalf("verdict %#v, want *Found pending with the bound intent", verdict)
	}
}

func TestReconcileNotFoundYetCarriesLoopState(t *testing.T) {
	client, mock := newMockClient(t, nil)
	mock.Reconcile("mock-mtx-0001").ReturnsNotFound()
	verdict, err := client.Reconcile(context.Background(), "mock-mtx-0001", policy(3))
	if err != nil {
		t.Fatal(err)
	}
	notFound, ok := verdict.(*revaly.NotFoundYet)
	if !ok {
		t.Fatalf("verdict %T, want *NotFoundYet", verdict)
	}
	if notFound.Attempts != 3 || notFound.LastHTTPStatus != 404 || notFound.LastCorrelationID == "" {
		t.Fatalf("loop state %+v, want 3 attempts / 404 / correlation", notFound)
	}
}

// Degraded reads (5xx and wire failures on the GET) keep polling within the
// budget — exactly the window where visibility is widest.
func TestReconcileKeepsPollingThroughDegradedReads(t *testing.T) {
	client, mock := newMockClient(t, nil)
	mock.Reconcile("mock-mtx-0001").Returns500().FailsBeforeSend().ReturnsFoundApproved()
	verdict, err := client.Reconcile(context.Background(), "mock-mtx-0001", policy(5))
	if err != nil {
		t.Fatal(err)
	}
	if found, ok := verdict.(*revaly.Found); !ok || found.Outcome != revaly.TransactionOutcomeApproved {
		t.Fatalf("verdict %#v, want *Found approved after degraded reads", verdict)
	}
	if len(mock.Requests()) != 3 {
		t.Fatalf("made %d attempts, want 3", len(mock.Requests()))
	}
}

func TestReconcile404ThenFound(t *testing.T) {
	client, mock := newMockClient(t, nil)
	mock.Reconcile("mock-mtx-0001").ReturnsNotFound().ReturnsFoundApproved()
	verdict, err := client.Reconcile(context.Background(), "mock-mtx-0001", policy(3))
	if err != nil {
		t.Fatal(err)
	}
	if found, ok := verdict.(*revaly.Found); !ok || found.Outcome != revaly.TransactionOutcomeApproved {
		t.Fatalf("verdict %#v, want *Found approved on the second poll", verdict)
	}
}

// A rejected READ that polling can never fix (non-404 PermanentRejection)
// escapes as the typed error.
func TestReconcileRejectedReadEscapes(t *testing.T) {
	client, mock := newMockClient(t, nil)
	mock.Reconcile("mock-mtx-0001").ReturnsPermanentRejection(401)
	verdict, err := client.Reconcile(context.Background(), "mock-mtx-0001", policy(5))
	var rejection *revaly.PermanentRejection
	if !errors.As(err, &rejection) || verdict != nil {
		t.Fatalf("got verdict=%v err=%v, want the 401 PermanentRejection escape", verdict, err)
	}
	if len(mock.Requests()) != 1 {
		t.Fatalf("made %d attempts, want 1 (no polling on rejected reads)", len(mock.Requests()))
	}
}

// A 2xx body this SDK cannot read at all is NOT a sighting — the loop
// continues (an ambiguous read licenses nothing).
func TestReconcileUnreadable2xxContinues(t *testing.T) {
	client, mock := newMockClient(t, nil)
	mock.Reconcile("mock-mtx-0001").ReturnsStatus(200, `[1,2,3]`).ReturnsFoundApproved()
	verdict, err := client.Reconcile(context.Background(), "mock-mtx-0001", policy(3))
	if err != nil {
		t.Fatal(err)
	}
	if found, ok := verdict.(*revaly.Found); !ok || found.Outcome != revaly.TransactionOutcomeApproved {
		t.Fatalf("verdict %#v, want *Found approved after the unreadable body", verdict)
	}
}

// Sightings this SDK version cannot bind are still FOUND (conservative:
// found-but-unmapped licenses no failover either way).
func TestReconcileUnmappedSightingsAreFound(t *testing.T) {
	for name, body := range map[string]string{
		"group envelope": `{"transaction":` + raptest.SyntheticTransaction(1) + `,"transactions":[` + raptest.SyntheticTransaction(1) + `]}`,
		"pending newer than the spec": `{"state":"pending","merchantTransactionId":"mock-mtx-0001",` +
			`"reservationExpiresAt":"2026-08-01T00:00:00Z"}`,
	} {
		client, mock := newMockClient(t, nil)
		mock.Reconcile("mock-mtx-0001").ReturnsStatus(200, body)
		verdict, err := client.Reconcile(context.Background(), "mock-mtx-0001", policy(2))
		if err != nil {
			t.Fatalf("%s: %v", name, err)
		}
		found, ok := verdict.(*revaly.Found)
		if !ok || found.Outcome != revaly.TransactionOutcomeUnknown {
			t.Fatalf("%s: verdict %#v, want *Found with outcome Unknown", name, verdict)
		}
	}
}

func TestReconcileHonorsContextCancellation(t *testing.T) {
	client, mock := newMockClient(t, nil)
	mock.Reconcile("mock-mtx-0001").ReturnsNotFound()
	ctx, cancel := context.WithCancel(context.Background())
	go func() {
		time.Sleep(50 * time.Millisecond)
		cancel()
	}()
	verdict, err := client.Reconcile(ctx, "mock-mtx-0001", revaly.ReconcilePolicy{
		MaxAttempts:   1000,
		OverallBudget: time.Hour,
		InitialDelay:  20 * time.Millisecond,
	})
	if verdict != nil || !errors.Is(err, context.Canceled) {
		t.Fatalf("got verdict=%v err=%v, want the context cancellation", verdict, err)
	}
}

func TestReconcilePolicyValidation(t *testing.T) {
	client, _ := newMockClient(t, nil)
	for name, bad := range map[string]revaly.ReconcilePolicy{
		"zero attempts": {MaxAttempts: 0, OverallBudget: time.Second},
		"zero budget":   {MaxAttempts: 1},
		"negative wait": {MaxAttempts: 1, OverallBudget: time.Second, InitialDelay: -time.Second},
	} {
		if _, err := client.Reconcile(context.Background(), "mock-mtx-0001", bad); err == nil {
			t.Fatalf("%s: expected a validation error", name)
		}
	}
	if _, err := client.Reconcile(context.Background(), "  ", policy(1)); err == nil {
		t.Fatal("blank merchantTransactionID: expected a validation error")
	}
}

// The budget bounds wall-clock spend: with a sticky 404 and a delay schedule
// that exceeds the budget, the loop stops early with NotFoundYet.
func TestReconcileBudgetBoundsTheLoop(t *testing.T) {
	client, mock := newMockClient(t, nil)
	mock.Reconcile("mock-mtx-0001").ReturnsNotFound()
	start := time.Now()
	verdict, err := client.Reconcile(context.Background(), "mock-mtx-0001", revaly.ReconcilePolicy{
		MaxAttempts:   50,
		OverallBudget: 150 * time.Millisecond,
		InitialDelay:  100 * time.Millisecond,
	})
	if err != nil {
		t.Fatal(err)
	}
	notFound, ok := verdict.(*revaly.NotFoundYet)
	if !ok {
		t.Fatalf("verdict %T, want *NotFoundYet", verdict)
	}
	if notFound.Attempts >= 50 {
		t.Fatalf("attempts = %d, want the budget to stop the loop early", notFound.Attempts)
	}
	if elapsed := time.Since(start); elapsed > 2*time.Second {
		t.Fatalf("loop ran %v, want it bounded by the budget", elapsed)
	}
}

File: languages/typescript/tests/reconcile.test.ts
import { describe, expect, test } from 'vitest';
import type { RapReconcileVerdict, ReconcilePolicy } from '../runtime/src/index';
import { RapMockTransport, RapPermanentRejection, SyntheticData } from '../runtime/src/index';
import { CollectingLogger } from './support/CollectingLogger';
import { mockedClient } from './support/TestClients';

/**
 * The failover-contract §3 procedure: both V1 verdicts, the poll-continue semantics of
 * degraded reads, the PermanentRejection escape, and the caller-bounded loop — the
 * only loop this SDK owns (ADR-SDK-004).
 */

const MTX = 'mtx-synthetic-1';

function immediatePolicy(maxAttempts = 1): ReconcilePolicy {
    return { maxAttempts, overallBudgetMs: 60_000, initialDelayMs: 0 };
}

describe('Found verdicts', () => {
    test.each([
        [1, 'Approved'],
        [2, 'Declined'],
        [3, 'Error'],
        [99, 'Unknown'],
    ])('transactionStatus %d maps to outcome %s', async (status, outcome) => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).returnsUnmappedStatus(status);

        const verdict = await mockedClient(mock).reconcile(MTX, immediatePolicy());

        expect(verdict.kind).toBe('Found');
        if (verdict.kind === 'Found') {
            expect(verdict.outcome).toBe(outcome);
            expect(verdict.transaction?.merchantTransactionId).toBe(MTX);
            expect(verdict.correlationId).toBe(SyntheticData.DEFAULT_CORRELATION_ID);
        }
    });

    test('a pending intent (post-P-2 shape) is Found(Pending) with the pending record', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).pending();

        const verdict = await mockedClient(mock).reconcile(MTX, immediatePolicy());

        expect(verdict.kind).toBe('Found');
        if (verdict.kind === 'Found') {
            expect(verdict.outcome).toBe('Pending');
            expect(verdict.pending?.state).toBe('pending');
            expect(verdict.pending?.merchantTransactionId).toBe(MTX);
            expect(verdict.transaction).toBeUndefined();
        }
    });

    test('a grouped envelope is Found(Unknown) — found-but-unmapped is still FOUND', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).returnsTransactionGroup();

        const verdict = await mockedClient(mock).reconcile(MTX, immediatePolicy());

        expect(verdict.kind).toBe('Found');
        if (verdict.kind === 'Found') {
            expect(verdict.outcome).toBe('Unknown');
        }
    });

    test('a state-bearing record this SDK cannot bind is still a sighting (Found Unknown)', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).returns(200, JSON.stringify({ state: 'reserved-for-p2' }));

        const verdict = await mockedClient(mock).reconcile(MTX, immediatePolicy());

        expect(verdict.kind).toBe('Found');
        if (verdict.kind === 'Found') {
            expect(verdict.outcome).toBe('Unknown');
        }
    });
});

describe('NotFoundYet and the poll loop', () => {
    test('404s then a terminal record: polls until Found', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).notFoundYet(2).thenFoundApproved();

        const verdict = await mockedClient(mock).reconcile(MTX, {
            maxAttempts: 5,
            overallBudgetMs: 60_000,
            initialDelayMs: 1,
        });

        expect(verdict.kind).toBe('Found');
        expect(mock.requests).toHaveLength(3);
    });

    test('NotFoundYet carries attempts, elapsed, last correlation id and last status', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).notFoundYet(3);

        const verdict = await mockedClient(mock).reconcile(MTX, {
            maxAttempts: 3,
            overallBudgetMs: 60_000,
            initialDelayMs: 0,
        });

        expect(verdict.kind).toBe('NotFoundYet');
        if (verdict.kind === 'NotFoundYet') {
            expect(verdict.attempts).toBe(3);
            expect(verdict.elapsedMs).toBeGreaterThanOrEqual(0);
            expect(verdict.lastHttpStatus).toBe(404);
            expect(verdict.lastCorrelationId).toBe(SyntheticData.DEFAULT_CORRELATION_ID);
        }
        expect(mock.requests).toHaveLength(3);
    });

    test('the overall budget bounds the loop even with attempts remaining', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).notFoundYet(1);

        const verdict = await mockedClient(mock).reconcile(MTX, {
            maxAttempts: 50,
            overallBudgetMs: 30,
            initialDelayMs: 100,
        });

        // The first wait (100ms) would already exceed the 30ms budget: stop after one attempt.
        expect(verdict.kind).toBe('NotFoundYet');
        if (verdict.kind === 'NotFoundYet') {
            expect(verdict.attempts).toBe(1);
        }
    });

    test('degraded reads (5xx on the GET) keep polling within the budget', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).returnsServerError().thenFoundApproved();

        const verdict = await mockedClient(mock).reconcile(MTX, immediatePolicy(3));

        expect(verdict.kind).toBe('Found');
        expect(mock.requests).toHaveLength(2);
    });

    test('transport failures on the GET keep polling within the budget', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).throwsConnectionRefused().thenFoundApproved();

        const verdict = await mockedClient(mock).reconcile(MTX, immediatePolicy(3));

        expect(verdict.kind).toBe('Found');
    });

    test('a 2xx with an unreadable body is an ambiguous read — poll again', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).returns(200, 'not json').thenFoundApproved();

        const verdict = await mockedClient(mock).reconcile(MTX, immediatePolicy(3));

        expect(verdict.kind).toBe('Found');
        expect(mock.requests).toHaveLength(2);
    });

    test('a rejected READ that polling can never fix escapes typed (401)', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).returnsPermanentRejection(401);

        await expect(mockedClient(mock).reconcile(MTX, immediatePolicy(5))).rejects.toBeInstanceOf(
            RapPermanentRejection,
        );
        expect(mock.requests).toHaveLength(1);
    });
});

describe('caller bounds and cancellation', () => {
    test('an empty merchantTransactionId is a caller error', async () => {
        await expect(mockedClient(new RapMockTransport()).reconcile('  ', immediatePolicy())).rejects.toThrow(
            'merchantTransactionId is required',
        );
    });

    test.each([
        [{ maxAttempts: 0, overallBudgetMs: 1000, initialDelayMs: 0 }, 'maxAttempts'],
        [{ maxAttempts: 1, overallBudgetMs: 0, initialDelayMs: 0 }, 'overallBudgetMs'],
        [{ maxAttempts: 1, overallBudgetMs: 1000, initialDelayMs: -1 }, 'initialDelayMs'],
    ])('invalid policy %j is rejected', async (policy, messagePart) => {
        await expect(mockedClient(new RapMockTransport()).reconcile(MTX, policy)).rejects.toThrow(messagePart);
    });

    test('aborting during a wait cancels the loop with the abort reason', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).notFoundYet(10);
        const controller = new AbortController();
        setTimeout(() => controller.abort(new Error('checkout gave up')), 20);

        await expect(
            mockedClient(mock).reconcile(
                MTX,
                { maxAttempts: 10, overallBudgetMs: 60_000, initialDelayMs: 5_000 },
                { signal: controller.signal },
            ),
        ).rejects.toThrow('checkout gave up');
    });

    test('the loop dispatches exactly its attempts — no hidden extra requests', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).notFoundYet(4);

        await mockedClient(mock).reconcile(MTX, { maxAttempts: 4, overallBudgetMs: 60_000, initialDelayMs: 0 });

        expect(mock.requests).toHaveLength(4);
    });
});

describe('verdict handling pattern', () => {
    test('the union forces a default branch (open for extension, ADR-SDK-009)', async () => {
        const mock = new RapMockTransport();
        mock.reconcile(MTX).returnsApproved();
        const logger = new CollectingLogger();

        const verdict: RapReconcileVerdict = await mockedClient(mock, { logger }).reconcile(
            MTX,
            immediatePolicy(),
        );

        // The quickstart's switch shape: every merchant integration writes this default.
        let handled: string;
        switch (verdict.kind) {
            case 'Found':
                handled = `found:${verdict.outcome}`;
                break;
            case 'NotFoundYet':
                handled = 'hold';
                break;
            default:
                // Future verdicts (SafeToFailover arrives with P-2) land here.
                handled = 'escalate';
                break;
        }
        expect(handled).toBe('found:Approved');
    });
});
